flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/recommendation/ALS.scala

```diff
@@ -581,6 +637,16 @@ object ALS {
         val userXy = new ArrayBuffer[Array[Double]]()
         val numRatings = new ArrayBuffer[Int]()
 
+        var precomputedXtX: Array[Double] = null
+
+        override def open(config: Configuration): Unit = {
+          // retrieve broadcasted precomputed XtX if using implicit feedback
+          if (implicitPrefs) {
+            precomputedXtX = getRuntimeContext.getBroadcastVariable[Array[Double]]("XtX")
+              .iterator().next()
+          }
+        }
+
         override def coGroup(left: lang.Iterable[(Int, Int, Array[Array[Double]])],
```

Comment:
If I see it right, I did not change this line, it was in the original ALS implementation. However, I can't find any reason to use the Java `Iterable`.

There could be other minor things to refactor in the original ALS code, but I preferred to keep them as they were, not to break anything. Should I refactor some parts along the way when I extend an algorithm like this?